### PR TITLE
when a ModelEntityItem moves, also update its meta-render-item

### PR DESCRIPTION
**Pull request by sethalves**
- when a ModelEntityItem moves, also update its meta-render-item
- fix some logic errors in UpdateEntityOperator

### CODE REVIEW

**Comment by sethalves** [comment]
to test:
- place one avatar/interface [A] somewhere in the domain
- with another interface [B], rez a model well behind that avatar
- use edit.js (with interface B) to find the ID of the model
- open the console on interface B
- do something like this:

  Entities.editEntity("{1a7940b4-d5ff-4456-bef1-0d19d9184d37}", {position:{x:74,y:1.5,z:80}});

  to "warp" the model in front of avatar A

- with this PR, interface A will see the model.  without this PR, it wont.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd2a29aace35ac6a30c9be6a9ba854a7e5839dae available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79141e38698f38226ae266337f4700292ed87958 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79141e38698f38226ae266337f4700292ed87958 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 574709824e924cba5da16a1ab474f48ec1ad4830 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 09317f46e5bc76ca35038124fa4232984cbf1e77 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 09317f46e5bc76ca35038124fa4232984cbf1e77 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ceb6aa582f4902d8e2fa8a07b6b1cde25c6222f4 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by ZappoMan** [comment]
This fixes a pretty serious bug... although it's not a regression since the last beta release, it's important enough that we should hold off on a new RC until we get this bug verified and merged.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a5efd3b822af157b7aec480740bdadd64bfdbe6 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by sethalves** [comment]
I hate you gustavo

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a5efd3b822af157b7aec480740bdadd64bfdbe6 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a5efd3b822af157b7aec480740bdadd64bfdbe6 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg

**Comment by MissML** [comment]
#### Success 
tested with @Kevin @Nex-Pro 
warping box test behaved correctly.

File: libraries/entities/src/UpdateEntityOperator.cpp [hunk 1]
```diff
@@ -43,32 +42,29 @@ UpdateEntityOperator::UpdateEntityOperator(EntityTreePointer tree,
...
+    if (oldElementBestFit) {
         if (_wantDebug) {
-            qCDebug(entities) << "    **** UNUSUAL CASE ****  no changes, but not best fit... consider it a move.... **";
+            qCDebug(entities) << "    **** TYPICAL NO MOVE CASE **** oldElementBestFit:" << oldElementBestFit;
         }
     } else {
-        _foundOld = true;
-        _newEntityCube = _oldEntityCube;
-        _dontMove = true;
+        _oldEntityBox = _existingEntity->getElement()->getAACube();
```
Comment (ZappoMan) [inline libraries/entities/src/UpdateEntityOperator.cpp]: ??? this looks suspicious...  why are you updating `_oldEntityBox` when you already set it to... 

```
    _oldEntityCube = _existingEntity->getQueryAACube();
    _oldEntityBox = _oldEntityCube.clamp((float)-HALF_TREE_SCALE, (float)HALF_TREE_SCALE); // clamp to domain bounds
```

File: libraries/entities/src/UpdateEntityOperator.cpp [hunk 2]
```diff
@@ -43,32 +42,29 @@ UpdateEntityOperator::UpdateEntityOperator(EntityTreePointer tree,
...
-        _newEntityCube = _oldEntityCube;
-        _removeOld = true; // our properties are going to move us, so remember this for later processing
-
+    if (oldElementBestFit) {
         if (_wantDebug) {
-            qCDebug(entities) << "    **** UNUSUAL CASE ****  no changes, but not best fit... consider it a move.... **";
+            qCDebug(entities) << "    **** TYPICAL NO MOVE CASE **** oldElementBestFit:" << oldElementBestFit;
         }
     } else {
-        _foundOld = true;
```
Comment (ZappoMan) [inline libraries/entities/src/UpdateEntityOperator.cpp]: I realize you flipped the logic... but I don't understand why you deleted this...

In the `oldElementBestFit == true` case, the path to the old and the path to the new should be the same exact path, and so you can short circuit half the tree traversal logic by setting `_foundOld = true`

Did you find a problem with that logic? Is that why you removed it?

File: libraries/entities/src/UpdateEntityOperator.cpp (hunk 1 above)
Comment (sethalves) [inline libraries/entities/src/UpdateEntityOperator.cpp]: This is the case where, for whatever reason, the entity isn't in the element that its position and size would imply.  _removeOld gets set to true, so it will get removed from the element its actually in, and _oldEntityCube is set to where the thing should have been to mark that area dirty.

In this case, I'm not actually sure which path should be marked dirty.

File: libraries/entities/src/UpdateEntityOperator.cpp (hunk 2 above)
Comment (sethalves) [inline libraries/entities/src/UpdateEntityOperator.cpp]: if I set that to true, the path that the thing should have been in wouldn't be marked dirty.  Maybe we don't need it to be?

File: libraries/entities/src/UpdateEntityOperator.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/entities/src/UpdateEntityOperator.cpp]: I'm just saying that you already set `_oldEntityBox` and you're resetting what you already set... so, you're right about the logic (that the entity is in the wrong element, and so we need to mark the old and mark the new)... but I'm saying you don't need to reset this `_oldEntityBox` value.

This comment is unrelated to my other comment.

File: libraries/entities/src/UpdateEntityOperator.cpp (hunk 2 above)
Comment (ZappoMan) [inline libraries/entities/src/UpdateEntityOperator.cpp]: If the old box is the best fit... then the new box and the new path should be the same as the old box and the old path... so you only need to traverse the path as the "new branch"... of course this optimization really has only a minimal effect, it doesn't reduce tree traversal, because basically the path is the same path... but it reduces the code run in the the `UpdateEntityOperator::preRecursion()` method... so this is not a big deal.

**Comment by samcake** [comment]
I would tag as  Dev approved for what i was worried about :)

File: libraries/entities/src/UpdateEntityOperator.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/entities/src/UpdateEntityOperator.cpp]: sold - you are correct

**Comment by hifi-gustavo** [comment]
Build console output for commit: d77b3bf59a6edec6b557628c15e33cc83f3d8fa2 available at: https://gist.github.com/54b48b4060ff6e39e5baea0b1678733d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7884.dmg
